### PR TITLE
Refactor AEAD errors

### PR DESCRIPTION
**Pull request by keks**
This is a smaller refactor to make the API more accessible.

Until now, the `aead` had a single error type that was returned by all the functions. While this make writing the module easy, it means that when calling a function, the caller needs to handle all errors returned by any of the functions in the module. Usually this means that they have to guess which errors are never called.

This PR introduces several error types, such that the caller only needs to handle error variants that actually can occur. It also updates the HPKE code using the `aead` module to properly handle the errors and adds a bunch of doc-comments.

### CODE REVIEW

File: src/aead.rs [hunk 1]
```diff
@@ -17,17 +17,120 @@ use crate::hacl::chacha20_poly1305;
...
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            DecryptError::UnsupportedAlgorithm => UnsupportedAlgorithmError.fmt(f),
+            DecryptError::DecryptionFailed => write!(f, "Decryption failed."),
+            DecryptError::InternalError => write!(f, "Internal error."),
+        }
+    }
+}
+
+impl std::error::Error for UnsupportedAlgorithmError {}
```
[the remark this answers is not in the document]
Comment (keks) [reply src/aead.rs:111]: Oh, alright. `core` is fine, though, right?

File: src/aead.rs [hunk 2]
```diff
@@ -306,39 +409,49 @@ fn decrypt_32(
...
+    msg_ctxt: &mut [u8],
+    iv: Iv,
+    aad: &[u8],
+) -> Result<Tag, EncryptError> {
     aesgcm::encrypt_128(&key.0, msg_ctxt, iv.0, aad)
         .map_err(|e| match e {
-            aesgcm::Error::UnsupportedHardware => Error::UnsupportedAlgorithm,
-            _ => Error::EncryptionError,
+            aesgcm::Error::UnsupportedHardware => EncryptError::UnsupportedAlgorithm,
+            _ => EncryptError::InternalError,
```
[the remark this answers is not in the document]
Comment (keks) [reply src/aead.rs:427]: Huh, good point. I'll revisit the different error cases and try to distinguish those better, so the caller gets a telling error (without leaking side channels).

File: src/aead.rs [hunk 3]
```diff
@@ -348,16 +461,22 @@ fn aes_decrypt_128(
     iv: Iv,
     aad: &[u8],
     tag: &Tag,
-) -> Result<(), Error> {
+) -> Result<(), DecryptError> {
     aesgcm::decrypt_128(&key.0, ctxt_msg, iv.0, aad, &tag.0).map_err(|e| match e {
-        aesgcm::Error::UnsupportedHardware => Error::UnsupportedAlgorithm,
-        _ => Error::EncryptionError,
+        aesgcm::Error::UnsupportedHardware => DecryptError::UnsupportedAlgorithm,
+        _ => DecryptError::InternalError,
```
[the remark this answers is not in the document]
Comment (keks) [reply src/aead.rs:467]: Ah, I see. I'll try to understand the error cases here better, as well.

**Comment by keks** [comment]
Thanks for the feedback. I am working on the CI - the tests passed locally, because I hadn't tested with all features, I believe.

Regarding the motivation: when writing the libcrux-based provider for OpenMLS, I ran into many cases I called a function and got an error with way to many error variants to be able to meaningfully handle then (where "handle" can mean "convert to different error and return that").

In addition, many variants just have very unclear semantics. What is `EncryptionError` when calling `encrypt`, and what is it when calling `decrypt`? In the provider, I just didn't know what to do with a lot of these errors, and just doing `.map_err(|_| Error::DecryptError)` would just silently eat internal errors (cf. the `LibraryError` discussion).

A while ago I read [this blog post](https://sabrinajewson.org/blog/errors), which I think has some really good points of how to design error types in Rust. I suppose the catchphrase of that article is

> Error types should be located near to their unit of fallibility

and in this case, the unit of fallibility of "encryption" (as a concept, not the specific function) is different from that of "decryption" or "consider these bytes as a key".

File: src/aead.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply src/aead.rs:111]: `Error` is `std` only unfortunately.

**Comment by keks** [comment]
I think I addressed all the feedback now. I decided to add simple third error `enum` instead of the `ErrorKind` pattern, because the latter seems mostly helpful if you want to attach more data to the error, which isn't necessary here. Otherwise it just would have been a pointless layer of indirection (`struct Error{ kind: ErrorKind } enum ErrorKind { ... }`).

There is one more change I made to `aesgcm`, I hope that is fine. The only error returned by `EverCrypt_AEAD_create_in` is that the algorithm in not supported on the given hardware, so I change the returned error: https://github.com/cryspen/libcrux/pull/197/files#diff-4de11af9cf9fecfb22c0c3304eb56e7daffdcf2219586f87703d1acb63b7d3e0L105-R105

File: src/aead.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply src/aead.rs]: You could've left the `Error` implementations but guard them with `std`. But let's skip them for now. We can always add them later.

**Review by franziskuskiefer — APPROVED** [review]
Thanks! I like this better. Let's get it in.

File: src/aead.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (keks) [reply src/aead.rs]: Yes, that would also have worked. I figured the previous errors also didn't implement it, so at least we are not losing anything.

**Pull request merged**